build-system/tasks/clean.js

```diff
@@ -3,75 +3,80 @@
 const argv = require('minimist')(process.argv.slice(2));
 const del = require('del');
 const fs = require('fs-extra');
+const globby = require('globby');
 const path = require('path');
 const {cyan, yellow} = require('kleur/colors');
+const {ignoreFile, splitIgnoreListByHeader} = require('./check-ignore-list');
 const {log} = require('../common/logging');
 
 const ROOT_DIR = path.resolve(__dirname, '../../');
 
+/**
+ * @return {Promise<string[]>}
+ */
+async function getPathsToDelete() {
+  const [, below] = splitIgnoreListByHeader(
+    await fs.readFile(ignoreFile, 'utf8')
+  );
+  return (
+    below
+      .split('\n')
+      // Comments and empty lines
+      .filter((line) => line.trim().length > 0 && !line.startsWith('#'))
+      // Recursive globs
+      .map((line) => (line.startsWith('/') ? line.substr(1) : `**/${line}`))
+  );
+}
+
 /**
  * Cleans up various cache and output directories. Optionally cleans up inner
  * node_modules package directories, or excludes some directories from deletion.
  * @return {Promise<void>}
  */
 async function clean() {
-  const pathsToDelete = [
-    // Local cache directories
-    // Keep this list in sync with .gitignore, .eslintignore, and .prettierignore
-    '.babel-cache',
-    '.css-cache',
-    '.jss-cache',
-    '.pre-closure-cache',
-
-    // Output directories
-    // Keep this list in sync with .gitignore, .eslintignore, and .prettierignore
-    '.amp-dep-check',
-    'build',
-    'build-system/dist',
-    'build-system/server/new-server/transforms/dist',
-    'build-system/tasks/performance/cache',
-    'build-system/tasks/performance/results.json',
-    'dist',
-    'dist.3p',
-    'dist.tools',
-    'export',
-    'examples/storybook',
-    'extensions/**/dist',
-    'release',
-    'result-reports',
-    'test/coverage',
-    'test/coverage-e2e',
-    'validator/**/dist',
-    'validator/export',
+  const pathsToDeleteFromIgnore = [
+    ...(await getPathsToDelete()),
+    '!**/third_party',
+    '!**/node_modules',
```

Comment:
Why are these two directories special-cased? Could this break the `--include_subpackages` logic?

Reply:
There are two lists of patterns executed on separate `del` calls. `--include_subpackages` will affect `pathsFromArgv`, which lacks these exclusions.